network/tcp.go

```diff
@@ -0,0 +1,422 @@
+	am, err := NewNetworkPacket(obj)
+	if err != nil {
+		return fmt.Errorf("Error converting packet: %v\n", err)
+	}
+	log.Lvlf5("Message SEND => %+v", am)
+	var b []byte
+	b, err = am.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("Error marshaling  message: %s", err.Error())
+	}
+	return c.send(b)
+}
+
+// send takes care of sending this slice of bytes FULLY to the connection
+func (c *TCPConn) send(b []byte) error {
+	// First write the size
+	packetSize := Size(len(b))
+	if err := binary.Write(c.conn, globalOrder, packetSize); err != nil {
+		return err
+	}
+	// Then send everything through the connection
+	// Send chunk by chunk
+	var sent Size
+	for sent < packetSize {
+		length := packetSize - sent
+		if length > maxChunkSize {
+			length = maxChunkSize
+		}
+
+		// Sending 'length' bytes
+		log.Lvl4("Sending from", c.conn.LocalAddr(), "to", c.conn.RemoteAddr())
+		n, err := c.conn.Write(b[:length])
+		if err != nil {
+			log.Error("Couldn't write chunk starting at", sent, "size", length, err)
+			return handleError(err)
+		}
+		sent += Size(n)
+		log.Lvl5("Sent", sent, "out of", packetSize)
+
+		// bytes left to send
+		b = b[n:]
+	}
+	// update stats on the connection
+	c.updateTx(uint64(packetSize))
+	return nil
+}
+
+// Remote returns the name of the peer at the end point of
+// the connection
+func (c *TCPConn) Remote() Address {
+	return c.endpoint
+}
+
+// Local returns the local address and port
+func (c *TCPConn) Local() Address {
+	return NewTCPAddress(c.conn.LocalAddr().String())
+}
+
+// Type implements the Conn interface
+func (c *TCPConn) Type() ConnType {
+	return PlainTCP
+}
+
+// Close ... closes the connection
+func (c *TCPConn) Close() error {
+	c.closedMut.Lock()
+	defer c.closedMut.Unlock()
+	if c.closed == true {
+		return nil
+	}
+	err := c.conn.Close()
+	c.closed = true
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// handleError produces the higher layer error depending on the type
+// so user of the package can know what is the cause of the problem
+func handleError(err error) error {
+
+	if strings.Contains(err.Error(), "use of closed") || strings.Contains(err.Error(), "broken pipe") {
+		return ErrClosed
+	} else if strings.Contains(err.Error(), "canceled") {
+		return ErrCanceled
+	} else if err == io.EOF || strings.Contains(err.Error(), "EOF") {
+		return ErrEOF
+	}
+
+	netErr, ok := err.(net.Error)
+	if !ok {
+		return ErrUnknown
+	}
+	if netErr.Temporary() {
+		return ErrTemp
+	} else if netErr.Timeout() {
+		return ErrTimeout
+	}
+	return ErrUnknown
+}
+
+// TCPListener is the underlying implementation of
+// Host using Tcp as a communication channel
+type TCPListener struct {
+	// the underlying golang/net listener
+	listener net.Listener
+	// the close channel used to indicate to the listener we want to quit
+	quit chan bool
+	// quitListener is a channel to indicate to the closing function that the
+	// listener has actually really quit
+	quitListener  chan bool
```

Comment:
I'm pretty sure they will refuse the way we do it, but maybe we should put an request feature or something because I guess we're not alone in this case. Having a `Close` method on the listener interface would be really nice.